proposals/dom_create_port.md

```diff
@@ -0,0 +1,206 @@
+a content script without leaking the details of that content script to the main
+world.  (Though we don't consider the isolated world boundary a security
+boundary, it is useful isolation and can serve as a "first line of defense").
+
+### Known Consumers
+
+User script managers have expressed an interest in this API.  More broadly,
+this would help any extension that needs to communicate between the main world
+of a web page with a more trusted script.
+
+## Specification
+
+### Schema
+
+```
+declare namespace dom {
+  interface PortProperties {
+    world: ExecutionWorld;
+    worldId?: string;
+  };
+
+  interface MessagePort {
+    sendMessage(args?: any): void
+    onMessage(args?: any): void
+  };
+
+  export function createPort(
+      properties: PortProperties
+  ): MessagePort;
+}
+```
+
+### Behavior
+
+#### Creating a Port
+
+An extension can create a new message port using `dom.createPort()`.  This will
+create and return a new message port according to the specifies `properties`
+(see below).
+
+#### Port Properties and World ID
+
+A message port can only be used to communicate with a single other world.  This:
+* Avoids the "many-to-one" opener-listener behavior that existing message ports
+  (created by `runtime.connect()`) in extensions have.  This behavior has caused
+  increased complexity and developer confusion.  A given port will have only one
+  channel.
+* Protects against accidentally sharing the port with another world.
+
+The caller indicates which world it would like the port to be associated with
+by specifying the `world` property to the appropriate type of execution world
+and, if appropriate, the `worldId` (e.g. for specific user script worlds).
+
+#### Passing a Message Port
+
+A message port can be passed to another world (in order to establish a
+connection) by leveraging the new `dom.execute()` API.  `dom.execute()` will be
+expanded to have special serialization logic for the MessagePort type.
+
+#### Example Usage
+
+Script that executes in the main world:
+```js
+function getFoo() { ... }
+function getBar() { ... }
+// Other code dispatches 'foochanged' and 'barchanged' events.
+```
+
+Content script:
+```js
+// This function will execute in the main world.
+function setUpMainWorld(port) {
+  // This responds to a message received from the content script.
+  port.onMessage = (message) => {
+    if (message == 'getFoo') {
+      const foo = getFoo();
+      port.sendMessage({foo});
+    } else if (message == 'getBar') {
+      const bar = getBar();
+      port.sendMessage({bar});
+    }
+  };
+
+  // These notify the content script of changes.
+  addEventListener('foochanged', () => {
+    port.sendMessage({foo: getFoo()});
+  });
+  addEventListener('barchanged', () => {
+    port.sendMessage({bar: getBar()});
+  });
+}
+
+// The rest of this code executes in the content script world.
+const mainWorldPort = browser.dom.createPort({world: 'MAIN'});
```

Comment:
Considering the port needs to be passed to the main world anyway. What is the reason to require explicitly specifying the world?

Reply:
See above from Oliver's comment.  This is to make it explicit and obvious to both the browser and the developer that this port is implicitly bound to one world, and disallow any accidental passing to another world or multiple worlds.